services/storage/handlers/namespaces.js

```diff
@@ -0,0 +1,253 @@
+'use strict';
+
+const NamespaceMember = require('../models/namespace-member');
+const Namespace = require('../models/namespace');
+const { response, fork } = require('boltzmann');
+const authn = require('../decorators/authn');
+const Package = require('../models/package');
+const User = require('../models/user');
+
+module.exports = [
+  fork.get('/v1/namespaces', namespaces),
+  fork.get('/v1/namespaces/namespace/:namespace([^@]+)@:host/members', authn.optional(members)),
+  fork.post(
+    '/v1/namespaces/namespace/:namespace([^@]+)@:host/members/:invitee',
+    authn.required(findInvitee(canChangeNamespace(invite)))
+  ),
+  fork.del(
+    '/v1/namespaces/namespace/:namespace([^@]+)@:host/members/:invitee',
+    authn.required(findInvitee(canChangeNamespace(remove)))
+  ),
+  fork.get(
+    '/v1/namespaces/namespace/:namespace([^@]+)@:host/maintainerships',
+    authn.optional(findNamespace(maintainerships))
+  )
+];
+
+function findInvitee(next) {
+  return async (context, params) => {
+    const user = await User.objects
+      .get({
+        active: true,
+        name: params.invitee
+      })
+      .catch(User.objects.NotFound, () => null);
+
+    if (!context.invitee) {
+      return response.error.coded('members.invite.invitee_dne', 404)
+    }
+
+    context.invitee = user;
+    return next(context, params);
+  };
+}
+
+function findNamespace(next) {
+  return async (context, params) => {
+    const ns = await Namespace.objects
+      .get({
+        active: true,
+        name: params.namespace,
+        'host.name': params.host,
+        'host.active': true
+      })
+      .catch(Namespace.objects.NotFound, () => null);
+
+    if (ns === null) {
+      return response.error.coded('members.invite.namespace_dne', 404)
+    }
+
+    context.namespace = ns;
+    return next(context, params);
+  };
+}
+
+// This is identical to isNamespaceMember except for the parameters read.
+// This one pays attention to the host.
+function canChangeNamespace(next) {
+  return async (context, params) => {
+    const ns = await Namespace.objects
+      .get({
+        active: true,
+        name: params.namespace,
+        'host.active': true,
+        'host.name': params.host,
+        'namespace_members.active': true,
+        'namespace_members.user_id': context.user.id
+      })
+      .catch(Namespace.objects.NotFound, () => null);
```

Comment:
Could we add a log here? #202 took me a bit to track down :sweat_smile:

Reply:
This _should_ appear a lot noisier in the logs now after adding dev-mode traces to error responses.